File: 알고리즘/프로그래머스/파이썬/가장 먼 노드.py
from collections import deque
def solution(n, edge):
    visited = [-1 for _ in range(n+1)]
    graph = {i : [] for i in range(1,n+1)}
    for e in edge:
        graph[e[1]].append(e[0])
        graph[e[0]].append(e[1])
    start = 1
    cost = 0
    queue = deque([[start,cost]])
    
    while queue:
        q = queue.popleft()
        start = q[0]
        cost = q[1]
        visited[start] = cost
        cost += 1
        for node in graph[start]:
            if visited[node] == -1:
                visited[node] = cost
                queue.append([node,cost])
    return visited.count(max(visited))

File: 알고리즘/프로그래머스/자바/가장 먼 노드.java
import java.util.Map;
import java.util.HashMap;
import java.util.List;
import java.util.ArrayList;
import java.util.Queue;
import java.util.LinkedList;
class Solution {
    public int solution(int n, int[][] edge) {
        int max = 0;
        Map<Integer, Integer> count = new HashMap<>();
        Map<Integer, List<Integer>> edges = new HashMap<>();
        boolean[] visited = new boolean[n+1];
        Queue<int[]> q = new LinkedList<>();
        
        for(int i = 1; i <= n; i++)
            edges.put(i, new ArrayList<>());
        
        for(int i = 0; i < edge.length; i++){
            edges.get(edge[i][0]).add(edge[i][1]);
            edges.get(edge[i][1]).add(edge[i][0]);
        }
        
        visited[1] = true;
        q.add(new int[] {1,0}); // 1번 노드, 거리 0  
	      while(!q.isEmpty()) {
	       	int[] now = q.poll();
	    	  max = max < now[1] ? now[1] : max;
	       	count.put(now[1], count.getOrDefault(now[1], 0) + 1);
 	        for(int i = 0; i < edges.get(now[0]).size(); i++) {
        		if(!visited[edges.get(now[0]).get(i)]) {
        			visited[edges.get(now[0]).get(i)] = true;
        			q.add(new int[] {edges.get(now[0]).get(i), now[1] + 1});
                }
	        }    	
	      }
        return count.get(max);
    }
}
